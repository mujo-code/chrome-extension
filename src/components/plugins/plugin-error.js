import React from 'react'
import { tracker } from '../../lib/error-tracker'

export class PluginError extends React.Component {
  constructor(props) {
    super(props)
    this.state = { hasError: false }
  }

  static getDerivedStateFromError(error) {
    return { hasError: true, error }
  }

  componentDidCatch(error, errorInfo) {
    tracker.exceptionWithInfo(error, errorInfo).then(errorId => {
      this.setState({ error, errorId })
      console.error(error, errorId)
    })
  }

  render() {
    const { hasError } = this.state
    if (hasError) {
      return <></>
    }

    return <>{this.props.children}</>
  }
}
