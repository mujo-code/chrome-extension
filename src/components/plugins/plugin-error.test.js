import React from 'react'
import { create, act } from 'react-test-renderer'
import { PluginError } from './plugin-error'

jest.mock('../../lib/error-tracker')
// eslint-disable-next-line import-order-alphabetical/order
const { tracker } = require('../../lib/error-tracker')

test('PluginError should match snapshot', () => {
  const tree = create(<PluginError>Foo</PluginError>).toJSON()
  expect(tree).toMatchSnapshot()
})

test('PluginError should match when there is an error', () => {
  tracker.exceptionWithInfo.mockReset().mockResolvedValue('foo')
  console.error = jest.fn()
  const ErrorComponent = () => {
    throw new Error('foo')
  }
  const wrapper = create(
    <PluginError>
      <ErrorComponent />
    </PluginError>
  )
  act(() => {
    wrapper.update()
  })
  expect(wrapper.toJSON()).toMatchSnapshot()
})

test('PluginError should track errors', () => {
  tracker.exceptionWithInfo.mockReset().mockResolvedValue('foo')
  console.error = jest.fn()
  const ErrorComponent = () => {
    throw new Error('foo')
  }
  const wrapper = create(
    <PluginError>
      <ErrorComponent />
    </PluginError>
  )
  act(() => {
    wrapper.update()
  })
  expect(tracker.exceptionWithInfo).toBeCalled()
  expect(console.error).toBeCalled()
})
