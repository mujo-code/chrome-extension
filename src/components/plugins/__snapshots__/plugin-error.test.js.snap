// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PluginError should match snapshot 1`] = `"Foo"`;

exports[`PluginError should match when there is an error 1`] = `null`;
